### PR TITLE
Async Crypto + Less magic to 'run in the browser'

### PR DESCRIPTION
**Pull request by dignifiedquire**
These are the changes that needed to bubble up from deep down in libp2p-crypto and `multihashing`.

This reduces the size of the browser build down to `2.4M` and `1.2M` unminified and minified respectively.

Associated PRs that need to be merged and shipped before this can get merged are

| PR | Next Aegir | CR | Sauce Credentials | RFM | M |
| --- | --- | --- | --- | --- | --- |
| https://github.com/libp2p/js-libp2p-crypto/pull/12 | ✔️ | ✔️ | ✔️  | ✔️ |✔️ |
| https://github.com/multiformats/js-multihashing-async/pull/1 | ✔️ | ✔️ | ✔️ |✔️  |  ✔️ |
| https://github.com/libp2p/js-peer-id/issues/33 | ✔️ | ✔️ | ✔️ |  ✔️ |✔️|
| https://github.com/libp2p/js-peer-info/issues/22 | ✔️ | ✔️ | ✔️  | ✔️  | ✔️ |
| https://github.com/libp2p/js-libp2p-identify/issues/6 | ✔️ | ✔️ | ✔️ |  ✔️ | ✔️ |
| https://github.com/libp2p/js-libp2p-swarm/issues/108 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/libp2p/js-libp2p-secio/issues/19 | ✔️ | ✔️ | ✔️ |  ✔️  | ✔️ |
| https://github.com/libp2p/js-libp2p-webrtc-star/pull/27 | ✔️ | ✔️ |  ✔️  | ✔️|✔️ |
| https://github.com/ipfs/js-libp2p-ipfs/pull/24 | ✔️ | ✔️ | N/A |✔️  | ✔️ |
| https://github.com/ipfs/js-libp2p-ipfs-browser/pull/99 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/ipfs/js-ipfs-block/pull/4 | ✔️ | ✔️ |✔️  | ✔️ | ✔️ |
| https://github.com/ipfs/js-ipfs-repo/pull/88 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/ipfs/interface-ipfs-core/pull/82 | ✔️  | ✔️ | N/A | ✔️ | ✔️ |
| https://github.com/ipfs/js-ipfs-api/pull/390 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/ipld/js-ipld-dag-pb/pull/3 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/ipld/js-ipld-dag-cbor/pull/28 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/ipfs/js-ipfs-unixfs/pull/12 | ✔️ |✔️  |✔️  | ✔️ | ✔️ |
| https://github.com/ipfs/js-ipfs-bitswap/pull/41 |✔️  | ✔️  | ✔️ | ✔️ | ✔️ |
| https://github.com/ipfs/js-cid/pull/5 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/ipfs/js-ipfs-block-service/pull/35 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/libp2p/interface-connection/pull/19 | ✔️ | ✔️ | N/A |✔️  | ✔️ |
| https://github.com/libp2p/js-libp2p-spdy/pull/36 | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| https://github.com/libp2p/js-libp2p-tcp/pull/40 | ✔️ | ✔️ | N/A | ✔️  | ✔️ |
| https://github.com/ipfs/js-ipfs-unixfs-engine/pull/82 | ✔️ | ✔️  | ✔️ | ✔️  | ✔️ |
| https://github.com/libp2p/js-libp2p-websockets/pull/33 | ✔️ | ✔️ |✔️ |✔️ | ✔️ |
| https://github.com/ipld/js-ipld-resolver/pull/65 | ✔️ | ✔️  | ✔️ | ✔️  | ✔️ |
| https://github.com/multiformats/js-multistream-select/pull/25 |✔️ | ✔️  |✔️  | ✔️ | ✔️ |
| #485 | ✔️ | ✔️  |✔️  | ✔️ | |
| https://github.com/dignifiedquire/aegir/pull/66 | ✔️ | ✔️ | N/A | ✔️ | ✔️ |

### TODOs
- [x] Update all PRs to use the new `multihashing-async` instead of `multihashing`
- [x] Use strings instead of buffers for protos, ref https://github.com/ipfs/js-ipfs-unixfs/pull/12#discussion_r81580704
- [x] Release new aegir with changes from https://github.com/ipfs/js-ipfs/issues/536
- [x] Update to use new aegir
- [x] Rebase onto IPLD changes

### CODE REVIEW

**Comment by diasdavid** [comment]
awesome! :D 

remember: if > half a day :)

**Comment by dignifiedquire** [comment]
@diasdavid ready for review

**Comment by diasdavid** [comment]
@dignifiedquire will defer my review until there are good benchmarks. Thank you for pushing this! :)

**Comment by dignifiedquire** [comment]
Doing more work on this, so not ready anymore ;)

**Comment by dignifiedquire** [comment]
@diasdavid tests are passing locally for node, so you can start reviewing :)

**Comment by dignifiedquire** [comment]
Added new todo

**Comment by nikuda** [comment]
Should `ipfs-bitswap` be on the above list too?

https://github.com/ipfs/js-ipfs-bitswap/tree/webcrypto

**Comment by dignifiedquire** [comment]
@nikuda yes it should :D

**Comment by nikuda** [comment]
Buffer to string changes needed in:

- [ ] libp2p/js-libp2p-crypto 
- [ ] libp2p/js-libp2p-secio
- [ ] ipfs/js-ipfs-bitswap
- [ ] ipfs/js-ipfs-merkle-dag
- [ ] ipfs/js-ipfs-unixfs

**Comment by dignifiedquire** [comment]
@nikuda awesome, thank you

**Comment by diasdavid** [comment]
From IRC

```
 <dignifiedquire> daviddias: good that you didn't review today :D I realised earlier that I hadn't tested the full down interop and had to fix some more issues, but now all is good and 💯 interop is back
```

Glad that I mentioned that several times :) Have you added CI tests to guarantee that interop remains? What were the tests you performed to verify?

**Comment by dignifiedquire** [comment]
> Glad that I mentioned that several times :) Have you added CI tests to guarantee that interop remains? What were the tests you performed to verify?

This interop are the tests you wanted to create scripts for. I just start a go and js node manually and connect them manually. 

There are a lot of tests to verify that we interop with go on libp2p-crypto already if you look at the tests there. But for the full integration test we need a full go-ipfs node from `master` that works

**Comment by diasdavid** [comment]
@dignifiedquire reviewed a bunch of this PR. ipfs-block needs some changes that will cascade to to the remaining PRs that I haven't finalized reviewing (i.e removed the 'needs review label')

**Comment by diasdavid** [comment]
Important to handle before moving forward with this PR - https://github.com/libp2p/js-libp2p-crypto/issues/18

**Comment by diasdavid** [comment]
Let's get this done! Reviewing all of the ones that have the first row with a green arrow :)

**Comment by diasdavid** [comment]
Let's have this meeeerged, pleeeease 🙏🏽

**Comment by diasdavid** [comment]
It says on the table that https://github.com/libp2p/js-libp2p-webrtc-star/pulls has a PR for next aegir, but there are not PR open

**Comment by dignifiedquire** [comment]
> It says on the table that https://github.com/libp2p/js-libp2p-webrtc-star/pulls has a PR for next aegir, but there are not PR open

fixed

**Comment by diasdavid** [comment]
The list is missing: https://github.com/multiformats/js-multistream-select/

**Comment by dignifiedquire** [comment]
> The list is missing: https://github.com/multiformats/js-multistream-select/

@diasdavid fixed

**Comment by diasdavid** [comment]
Ok, let's start :)

**Comment by diasdavid** [comment]
https://github.com/libp2p/interface-stream-muxer was missing, I've updated it to aegir 9 :)

**Comment by diasdavid** [comment]
Ok status report:

- We are pretty close
- Seeing some errors on js-ipfs-api with the new 'fixtures' thing
- Some modules fail to authenticate in Saucelabs (e.g peer-id, peer-info) and so travis fails
- libp2p-ipfs-browser fails on tests that pass flawlessly in node.js, but not in the browser

**Comment by dignifiedquire** [comment]
and another one: https://github.com/ipfs/interface-ipfs-core/pull/84

File: package.json [hunk 1]
```diff
@@ -5,21 +5,28 @@
...
     "jsipfs": "src/cli/bin.js"
   },
-  "main": "lib/core/index.js",
-  "jsnext:main": "src/core/index.js",
+  "main": "src/core/index.js",
+  "browser": {
+    "libp2p-ipfs": "libp2p-ipfs-browser",
+    "./src/core/default-repo.js": "./src/core/default-repo-browser.js",
+    "./src/core/components/init-assets.js": false,
+    "fs-pull-blob-store": false,
```
Comment (diasdavid) [inline package.json]: does this one get pulled with line 11 happening?

**Comment by diasdavid** [comment]
re: this PR, so far looks good to me :) 👍

File: package.json (hunk 1 above)
Comment (dignifiedquire) [inline package.json]: good question, need to recheck if it is still needed

**Comment by dignifiedquire** [comment]
Another one: https://github.com/libp2p/js-peer-book/pull/12

**Comment by dignifiedquire** [comment]
Now that the changes have happened we need to update https://github.com/ipfs/community/blob/master/js-project-guidelines.md accordingly

**Comment by diasdavid** [comment]
Currently blocked by:

- https://github.com/ipfs/js-libp2p-ipfs-browser/pull/99
- https://github.com/libp2p/js-libp2p-secio/pull/30
- https://github.com/libp2p/js-libp2p-crypto/pull/29

**Comment by dignifiedquire** [comment]
Unblocked with https://github.com/libp2p/js-libp2p-crypto/pull/29

**Comment by dignifiedquire** [comment]
All dependencies were merged and released, doing the last updates now :tada:

**Comment by dignifiedquire** [comment]
@diasdavid can I get some last review on this?

**Comment by diasdavid** [comment]
@dignifiedquire all CI is failing, is that what you are expecting?

File: package.json (hunk 1 above)
Comment (diasdavid) [inline package.json]: Needs to be reviewed

File: src/cli/commands/block/get.js [hunk 2]
```diff
@@ -19,11 +19,11 @@ module.exports = {
         throw err
       }
 
-      const mh = utils.isDaemonOn()
+      const hash = utils.isDaemonOn()
         ? argv.key
-        : new Buffer(bs58.decode(argv.key))
+        : mh.fromB58String(argv.key)
```
Comment (diasdavid) [inline src/cli/commands/block/get.js]: Wait, does this still apply? we have standardised the block API

File: src/cli/commands/block/rm.js [hunk 3]
```diff
@@ -24,9 +24,7 @@ module.exports = {
         throw new Error('rm block with daemon running is not yet implemented')
       }
 
-      const mh = new Buffer(bs58.decode(argv.key))
-
-      ipfs.block.del(mh, (err) => {
+      ipfs.block.del(mh.fromB58String(argv.key), (err) => {
```
Comment (diasdavid) [inline src/cli/commands/block/rm.js]: block.del should support multihash as strings

File: src/cli/commands/object/get.js [hunk 4]
```diff
@@ -13,24 +14,17 @@ module.exports = {
   builder: {},
 
   handler (argv) {
-    utils.getIPFS((err, ipfs) => {
+    waterfall([
+      (cb) => utils.getIPFS(cb),
+      (ipfs, cb) => ipfs.object.get(argv.key, {enc: 'base58'}, cb),
+      (node, cb) => node.toJSON(cb)
+    ], (err, res) => {
```
Comment (diasdavid) [inline src/cli/commands/object/get.js]: s/res/nodeJSON

File: src/cli/commands/object/new.js [hunk 5]
```diff
@@ -13,23 +14,16 @@ module.exports = {
   builder: {},
 
   handler (argv) {
-    utils.getIPFS((err, ipfs) => {
+    waterfall([
+      (cb) => utils.getIPFS(cb),
+      (ipfs, cb) => ipfs.object.new(cb),
+      (node, cb) => node.toJSON(cb)
+    ], (err, node) => {
```
Comment (diasdavid) [inline src/cli/commands/object/new.js]: s/node/nodeJSON

File: src/cli/commands/object/patch/add-link.js [hunk 6]
```diff
@@ -16,73 +17,29 @@ module.exports = {
...
-          nodePatched = node
-          cb()
-        })
-      }
-    ], (err) => {
+        const link = new DAGLink(argv.name, stats[0], stats[1])
+        ipfs.object.patch.addLink(argv.root, link, {enc: 'base58'}, cb)
+      }),
+      (node, cb) => node.toJSON(cb)
+    ], (err, node) => {
```
Comment (diasdavid) [inline src/cli/commands/object/patch/add-link.js]: s/node/nodeJSON

File: src/cli/commands/object/patch/set-data.js [hunk 7]
```diff
@@ -3,29 +3,22 @@
...
 const log = debug('cli:object')
 log.error = debug('cli:object:error')
 
 function parseAndAddNode (key, data) {
-  utils.getIPFS((err, ipfs) => {
+  waterfall([
+    (cb) => utils.getIPFS(cb),
+    (ipfs, cb) => ipfs.object.patch.setData(key, data, {enc: 'base58'}, cb),
+    (node, cb) => node.toJSON(cb)
+  ], (err, node) => {
```
Comment (diasdavid) [inline src/cli/commands/object/patch/set-data.js]: s/node/nodeJSON

File: src/cli/commands/object/put.js [hunk 8]
```diff
@@ -3,29 +3,22 @@
...
 const log = debug('cli:object')
 log.error = debug('cli:object:error')
 
 function putNode (buf, enc) {
-  utils.getIPFS((err, ipfs) => {
+  waterfall([
+    (cb) => utils.getIPFS(cb),
+    (ipfs, cb) => ipfs.object.put(buf, {enc}, cb),
+    (node, cb) => node.toJSON(cb)
+  ], (err, node) => {
```
Comment (diasdavid) [inline src/cli/commands/object/put.js]: s/node/nodeJSON

File: src/cli/commands/object/put.js [hunk 9]
```diff
@@ -3,29 +3,22 @@
...
+const waterfall = require('async/waterfall')
 const debug = require('debug')
 const log = debug('cli:object')
 log.error = debug('cli:object:error')
 
 function putNode (buf, enc) {
-  utils.getIPFS((err, ipfs) => {
+  waterfall([
+    (cb) => utils.getIPFS(cb),
+    (ipfs, cb) => ipfs.object.put(buf, {enc}, cb),
```
Comment (diasdavid) [inline src/cli/commands/object/put.js]: { enc: enc }

File: src/cli/commands/object/put.js [hunk 10]
```diff
@@ -3,29 +3,22 @@
 const utils = require('../../utils')
 const bl = require('bl')
 const fs = require('fs')
+const waterfall = require('async/waterfall')
 const debug = require('debug')
 const log = debug('cli:object')
 log.error = debug('cli:object:error')
 
 function putNode (buf, enc) {
```
Comment (diasdavid) [inline src/cli/commands/object/put.js]: s/enc/encoding

File: src/core/components/init-assets.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    pull.map((element) => {
+      const addPath = element.substring(index + 1, element.length)
+      if (fs.statSync(element).isDirectory()) return
+
+      return {
+        path: addPath,
+        content: file(element)
+      }
+    }),
+    pull.filter(Boolean),
```
Comment (diasdavid) [inline src/core/components/init-assets.js]: What is this exactly?

File: src/core/components/init-assets.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+  pull(
+    pull.values([initDocsPath]),
+    pull.asyncMap((val, cb) => {
+      glob(path.join(val, '/**/*'), cb)
+    }),
+    pull.flatten(),
+    pull.map((element) => {
+      const addPath = element.substring(index + 1, element.length)
+      if (fs.statSync(element).isDirectory()) return
```
Comment (diasdavid) [inline src/core/components/init-assets.js]: curly

File: src/core/components/init-assets.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      }
+    }),
+    pull.filter(Boolean),
+    importer(self._ipldResolver),
+    pull.through((el) => {
+      if (el.path === 'files/init-docs/docs') {
+        const hash = mh.toB58String(el.multihash)
+        log('to get started, enter:')
+        log()
+        log(`\t jsipfs files cat /ipfs/${hash}/readme`)
```
Comment (diasdavid) [inline src/core/components/init-assets.js]: Let's point directly to the readme hash, since we don't do traversals right now //cc @VictorBjelkholm

File: src/core/default-repo/index.js [hunk 14]
```diff
@@ -1,7 +0,0 @@
-'use strict'
-
-const isNode = require('detect-node')
-
-module.exports = isNode
-  ? require('./node')
-  : require('./browser')
```
Comment (diasdavid) [inline src/core/default-repo/index.js]: super sweet that we can get webpack, browserify and others to pick up browser vs node files through package.json 👍

File: src/http-api/resources/block.js [hunk 15]
```diff
@@ -80,10 +81,17 @@ exports.put = {
   // main route handler which is called after the above `parseArgs`, but only if the args were valid
   handler: (request, reply) => {
     const data = request.pre.args.data
-
-    const block = new Block(data)
-
-    request.server.app.ipfs.block.put(block, (err) => {
+    const ipfs = request.server.app.ipfs
```
Comment (diasdavid) [inline src/http-api/resources/block.js]: thank you!

File: src/http-api/resources/object.js [hunk 16]
```diff
@@ -31,7 +33,11 @@ exports.parseKey = (request, reply) => {
 }
 
 exports.new = (request, reply) => {
-  request.server.app.ipfs.object.new((err, node) => {
+  const ipfs = request.server.app.ipfs
+  waterfall([
+    (cb) => ipfs.object.new(cb),
+    (node, cb) => node.toJSON(cb)
+  ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON

File: src/http-api/resources/object.js [hunk 17]
```diff
@@ -60,8 +58,12 @@ exports.get = {
   handler: (request, reply) => {
     const key = request.pre.args.key
     const enc = request.query.enc || 'base58'
+    const ipfs = request.server.app.ipfs
 
-    request.server.app.ipfs.object.get(key, {enc}, (err, node) => {
+    waterfall([
+      (cb) => ipfs.object.get(key, {enc}, cb),
+      (node, cb) => node.toJSON(cb)
+    ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON

File: src/http-api/resources/object.js [hunk 18]
```diff
@@ -156,10 +149,14 @@ exports.put = {
...
-
-    request.server.app.ipfs.object.put(node, (err, obj) => {
+    const node = request.pre.args.node
+    const dagNode = new DAGNode(new Buffer(node.Data), node.Links)
+    const ipfs = request.server.app.ipfs
+
+    parallel([
+      (cb) => ipfs.object.put(dagNode, cb),
+      (cb) => dagNode.toJSON(cb)
+    ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON

File: src/http-api/resources/object.js [hunk 19]
```diff
@@ -233,8 +221,12 @@ exports.links = {
   // main route handler which is called after the above `parseArgs`, but only if the args were valid
   handler: (request, reply) => {
     const key = request.pre.args.key
+    const ipfs = request.server.app.ipfs
 
-    request.server.app.ipfs.object.get(key, (err, node) => {
+    waterfall([
+      (cb) => ipfs.object.get(key, cb),
+      (node, cb) => node.toJSON(cb)
+    ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON

File: src/http-api/resources/object.js [hunk 20]
```diff
@@ -338,8 +318,12 @@ exports.patchSetData = {
   handler: (request, reply) => {
     const key = request.pre.args.key
     const data = request.pre.args.data
+    const ipfs = request.server.app.ipfs
 
-    request.server.app.ipfs.object.patch.setData(key, data, (err, node) => {
+    waterfall([
+      (cb) => ipfs.object.patch.setData(key, data, cb),
+      (node, cb) => node.toJSON(cb)
+    ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON/g

File: src/http-api/resources/object.js [hunk 21]
```diff
@@ -416,46 +393,26 @@ exports.patchAddLink = {
...
+        (cb) => parallel([
+          (cb) => linkedObj.size(cb),
+          (cb) => linkedObj.multihash(cb)
+        ], cb),
+        (stats, cb) => {
+          cb(null, new DAGLink(name, stats[0], stats[1]))
+        },
+        (link, cb) => ipfs.object.patch.addLink(root, link, cb),
+        (node, cb) => node.toJSON(cb)
+      ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON/g

File: src/http-api/resources/object.js [hunk 22]
```diff
@@ -493,8 +450,12 @@ exports.patchRmLink = {
   handler: (request, reply) => {
     const root = request.pre.args.root
     const link = request.pre.args.link
+    const ipfs = request.server.app.ipfs
 
-    request.server.app.ipfs.object.patch.rmLink(root, link, (err, node) => {
+    waterfall([
+      (cb) => ipfs.object.patch.rmLink(root, link, cb),
+      (node, cb) => node.toJSON(cb)
+    ], (err, res) => {
```
Comment (diasdavid) [inline src/http-api/resources/object.js]: s/res/nodeJSON

File: test/core/both/index.js [hunk 23]
```diff
@@ -1,14 +1,13 @@
...
-    require('./' + file)
-  })
+  require('./test-bitswap')
+  require('./test-block')
+  require('./test-bootstrap')
+  require('./test-config')
+  require('./test-files')
+  require('./test-generic')
+  require('./test-init')
+  require('./test-object')
```
Comment (diasdavid) [inline test/core/both/index.js]: 👍

File: src/cli/commands/object/put.js (hunk 10 above)
Comment (dignifiedquire) [inline src/cli/commands/object/put.js]: 👎  unrelated change

File: src/cli/commands/block/get.js (hunk 2 above)
Comment (dignifiedquire) [inline src/cli/commands/block/get.js]: yes, tests will fail otherwise

File: src/cli/commands/block/rm.js (hunk 3 above)
Comment (dignifiedquire) [inline src/cli/commands/block/rm.js]: that's exactly what I'm doing here, decoding the multihash from a string

File: package.json (hunk 1 above)
Comment (dignifiedquire) [inline package.json]: need it as it is used in the tests, and does no harm otherwise

**Comment by dignifiedquire** [comment]
> @dignifiedquire all CI is failing, is that what you are expecting?

yes, waiting for the bitswap PR I pinged you about

File: src/http-api/resources/object.js (hunk 22 above)
Comment (dignifiedquire) [inline src/http-api/resources/object.js]: Seems overly verbose, and I really dislike all caps naming. Not sure about a better name

File: src/core/components/init-assets.js (hunk 13 above)
Comment (dignifiedquire) [inline src/core/components/init-assets.js]: Let's not add more things to this large PR that we need to, that is an easy fix we can do afterwards

File: src/core/components/init-assets.js (hunk 11 above)
Comment (dignifiedquire) [inline src/core/components/init-assets.js]: It filters out all `undefined` elements, which are directories, if you check a couple of lines above

Comment (diasdavid) [inline src/core/components/init-assets.js]: could you add a comment saying that?

**Comment by diasdavid** [comment]
@dignifiedquire there you go https://github.com/ipfs/js-ipfs-bitswap/pull/67

File: src/http-api/resources/object.js (hunk 22 above)
Comment (diasdavid) [inline src/http-api/resources/object.js]: well, toJSON is also JSON all caps

File: src/core/components/init-assets.js (hunk 11 above)
Comment (dignifiedquire) [inline src/core/components/init-assets.js]: done

File: src/http-api/resources/object.js (hunk 22 above)
Comment (dignifiedquire) [inline src/http-api/resources/object.js]: because apis in JavaScript are note very consistent when it comes to naming :/

**Comment by dignifiedquire** [comment]
CI is finally green, browsers are not perfect atm, most are failing due to one or the other issue

Firefox fails with this
```
Firefox 49.0.0 (Linux 0.0.0) ERROR
  SyntaxError: missing = in const declaration
  at webpack:///~/cbor/lib/encoder.js:383:0 <- test/browser.js:41233
```

though this code runs fine in Chrome. 

@diasdavid if you are happy with the code please merge and ship at your earliest convenience, and I will investigate and fix various browser related issues next week.

**Comment by diasdavid** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉
